File: python/RandomForest.py
#!/usr/bin/env python3

import logging
import random

import CSVReader
from DecisionTree import DecisionTreeClassifier


"""
TODO: - add arguments to tree (max_depth, criterion)
      - add same arguments to random forest
      - implement a thread pool to fit multiple trees at the same time
      - comment functions
      - set code to PEP8 norm
      - unittests
"""


class RandomForestClassifier(object):

    def __init__(self, nb_trees, nb_samples):
        self.trees = []
        self.nb_trees = nb_trees
        self.nb_samples = nb_samples


    def fit(self, data):
        for i in range(self.nb_trees):
            logging.info('Training tree {}'.format(i + 1))

            random_features = random.sample(data, self.nb_samples)
            tree = DecisionTreeClassifier()
            tree.fit(random_features)

            self.trees.append(tree)


    def predict(self, feature):
        predictions = []

        for tree in self.trees:
            predictions.append(tree.predict(feature))

        return max(set(predictions), key=predictions.count)


def test_rf():
    # random_split the features & log error rate
    pass


if __name__=='__main__':
    logging.basicConfig(level=logging.INFO)

    data = CSVReader.read_csv("../scala/data/income.csv")
    rf = RandomForestClassifier(30, 1000)
    rf.fit(data)

    print(rf.predict([39, 'State-gov', 'Bachelors', 13, 'Never-married', \
                    'Adm-clerical', 'Not-in-family', 'White', 'Male', \
                    2174, 0, 40, 'United-States']))

File: scala/src/main/scala/random_forest.scala
package com.ensemble

import scala.util.Random
import scala.collection.mutable.{Map => MMap}
import com.tree.{DecisionTreeClassifier, Types}


class RandomForestClassifier(protected val nbTrees: Int,
                            protected val nbSamples: Int,
                            protected val maxDepth: Int = -1) {

  protected val trees = MMap[Int, DecisionTreeClassifier]()

  /**
   * Trains a random forest classifier using the iven dataset.
   *
   * @param features  List of Lists containing the input variables
   * @param targets   List of outputs corresponding to the given features
   */
  def fit(features: Types.Features, targets: List[String]) = {
    0 to nbTrees foreach { i =>
      println(s"Training tree ${i}")

      val randomSet = Random.shuffle(features.zip(targets)).take(nbSamples)
      val randomFeatures = randomSet map {tuple => tuple._1}
      val randomTargets = randomSet map {tuple => tuple._2}

      val tree = new DecisionTreeClassifier(maxDepth=maxDepth)
      tree.fit(randomFeatures, randomTargets)

      trees(i) = tree
    }
  }

  /**
   * Makes a prediction using the previously trained trees and the given
   * feature. It uses the "max votes" technique.
   *
   * @param feature A List containing the feature to make a prediction for
   * @return        The prediction (String)
   */
  def predict(feature: List[Any]) = {
    trees
      .values
      .map { tree => tree.predict(feature) }
      .groupBy(identity)
      .maxBy(_._2.size)
      ._1
  }

}
